async/deferred.ts

```diff
@@ -20,7 +21,17 @@ export interface Deferred<T> extends Promise<T> {
 export function deferred<T>(): Deferred<T> {
   let methods;
   const promise = new Promise<T>((resolve, reject): void => {
-    methods = { resolve, reject };
+    methods = {
+      resolve(value: T | PromiseLike<T>) {
+        Object.assign(promise, { status: "fulfilled" });
```

Comment:
The timing of setting "fulfilled" looks too early to me when the value is actually a promise.

I think these lines should be something like:
```ts
await value;
Object.assign(promise, { status: "fulfilled" });
resolve(value);
```

---
note: when I checked the following expression in deno repl:

```ts
let a = new Promise((resolve) => resolve(new Promise(() => {})));
```

then console.log(a) prints `Promise { <pending> }` forever. So a promise is fulfilled when the resolved value is fulfilled.

Reply:
Nice catch! 
Indeed this case was not handled properly, I applied your suggested changes and added another test to cover it 👍